src/engine/input/RunOnInputAttribute.cs

```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+public abstract class RunOnInputAttribute : Attribute
+{
+    ~RunOnInputAttribute()
```

Comment:
Are attributes ever destructed? If the assembly is unloaded does this run then?

Reply:
[https://stackoverflow.com/questions/3235302/when-is-the-desctructor-called-for-a-custom-attribute-class](yes, it gets called)

> Therefore a method which uses GetCustomAttributes() to retrieve an attribute, uses it and then throws away the reference has just released a new instance of that attribute for the GC to clean up when it needs to.

>Therefore - attribute instances are governed by the exact same memory management and lifetime rules as all class instances; therefore what PieterG says in his answer is correct - the destructor could be called at any time after all references to that attribute have been released.

Reply:
That seems like the classic GC language problem, it is not guaranteed that this is run at expected time.
If `AttributesWithMethods.RemoveAll(p => p.Item2.Equals(this));` is relied upon to be actually ran at some point, then it needs to be re-designed so that that code is ran consistently when needed.

Reply:
There is no relieance on this code to run, but It will speed up the foreach calls.

Reply:
I ran this with a debugger and, this gets triggered on game close, which I suppose happens because the Thrive.dll was unloaded. So fine, this can stay. Though maybe a comment should be added so that this same discussion doesn't need to be had in the future.

Reply:
I mean, if the desctructor only gets called when closing the game, I could also remove the `Dispose()`, because we don't really care about cleaning up the list after the game is closed

Reply:
I figured that, as you said, mods, could in theory be unloaded without closing the game, causing attributes from the mod to be destructed.

Reply:
ah, that makes sense